core/lib/spree/event.rb

```diff
@@ -52,51 +110,78 @@ def fire(event_name, opts = {})
     #   end
     #
     # @see Spree::Event#unsubscribe
-    def subscribe(event_name, &block)
-      name = normalize_name(event_name)
-      listener_names << name
-      adapter.subscribe(name, &block)
+    def subscribe(event_name, adapter: default_adapter, &block)
+      event_name = normalize_name(event_name)
+      adapter.subscribe(event_name, &block).tap do
+        if legacy_adapter(adapter)
+          listener_names << adapter.normalize_name(event_name)
+        end
+      end
     end
 
     # Unsubscribes a whole event or a specific subscription object
     #
-    # @param [String, Object] subscriber the event name as a string (with
-    #  or without the ".spree" suffix) or the subscription object
+    # When unsubscribing from an event, all previous listeners are deactivated.
+    # Still, you can add new subscriptions to the same event and they'll be
+    # called if the event is fired:
+    #
+    # @example
+    #   Spree::Event.subscribe('foo') { do_something }
+    #   Spree::Event.unsubscribe 'foo'
+    #   Spree::Event.subscribe('foo') { do_something_else }
+    #   Spree::Event.fire 'foo' # `do_something_else` will be called, but
+    #   # `do_something` won't
+    #
+    # @param [String, Symbol, Spree::Event::Listener] name_or_subscriber the
+    # event name as a string or the subscription object. Take into account that
+    # if the deprecated {Spree::Event::Adapters::ActiveSupportNotifications}
+    # adapter is used, the subscription object will be a
+    # {ActiveSupport::Notifications::Fanout::Subscribers::Timed} object.
     #
     # @example Unsubscribe a single subscription
-    #   subscription = Spree::Event.fire 'order_finalized'
+    #   subscription = Spree::Event.subscribe('order_finalized') { do_something
+    #   }
     #   Spree::Event.unsubscribe(subscription)
     # @example Unsubscribe all `order_finalized` event subscriptions
     #   Spree::Event.unsubscribe('order_finalized')
-    # @example Unsubscribe an event by name with explicit prefix
-    #   Spree::Event.unsubscribe('order_finalized.spree')
-    def unsubscribe(subscriber)
-      name_or_subscriber = subscriber.is_a?(String) ? normalize_name(subscriber) : subscriber
+    def unsubscribe(name_or_subscriber, adapter: default_adapter)
+      name_or_subscriber = normalize_name(name_or_subscriber)
```

Comment:
I don't get how `normalize_name` is handling the case when `name_or_subscriber` is a `subscriber`. Will that just be a symbol in that case? In the method documentation I see also `Spree::Event::Listener`

Reply:
`normalize_name` is `Spree::Event` is just returning the argument untouched unless it's a Symbol. So, if it's a `Spree::Event::Listener` it won't do anything. I agree it'd be clearer to ask whether it's a `Listener` before dispatching it to `normalize_name`, but as we can be dealing with an `ActiveSupport...Fanout` object from the legacy adapter it'd make the code rather ugly. WDYT?

Reply:
My only concern is that looking at the `normalize_name` method as is, I assume that name can be a String or a Symbol, even if as you pointed out, it can be any object. Not sure if any action is needed honestly, the code looks good, maybe it's just me. :)

Reply:
I think you're right. We can't let the legacy adapter condition a better design. I refactored the method so that `normalize_name` is only called when it's not a subscriber object from any of the adapters.